src/Microsoft.Identity.Client/AppConfig/AbstractApplicationBuilder.cs

```diff
@@ -0,0 +1,412 @@
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <returns></returns>
+        public T WithDebugLoggingCallback()
+        {
+            if (Config.LoggingCallback != null)
+            {
+                throw new InvalidOperationException("LoggingCallback has already been set");
+            }
+
+            Config.LoggingCallback = (level, message, pii) => { Debug.WriteLine($"{level}: {message}"); };
+            return (T)this;
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <returns></returns>
+        public T WithTelemetryHandler(ITelemetryHandler telemetryHandler)
+        {
+            if (Config.TelemetryHandler != null)
+            {
+                throw new InvalidOperationException("TelemetryHandler has already been set");
+            }
+
+            Config.TelemetryHandler = telemetryHandler;
+            return (T)this;
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="clientId"></param>
+        /// <returns></returns>
+        protected T WithClientId(string clientId)
+        {
+            if (!string.IsNullOrWhiteSpace(clientId))
+            {
+                Config.ClientId = clientId;
+            }
+
+            return (T)this;
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="redirectUri"></param>
+        /// <returns></returns>
+        public T WithRedirectUri(string redirectUri)
+        {
+            Config.RedirectUri = GetValueIfNotEmpty(Config.RedirectUri, redirectUri);
+            return (T)this;
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="tenant"></param>
+        /// <returns></returns>
+        public T WithTenant(string tenant)
+        {
+            Config.Tenant = GetValueIfNotEmpty(Config.Tenant, tenant);
+            return (T)this;
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="enablePiiLogging"></param>
+        /// <returns></returns>
+        public T WithEnablePiiLogging(bool enablePiiLogging)
+        {
+            Config.EnablePiiLogging = enablePiiLogging;
+            return (T)this;
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="logLevel"></param>
+        /// <returns></returns>
+        public T WithLoggingLevel(LogLevel logLevel)
+        {
+            Config.LogLevel = logLevel;
+            return (T)this;
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="enabled"></param>
+        /// <returns></returns>
+        public T WithDefaultPlatformLoggingEnabled(bool enabled)
+        {
+            Config.IsDefaultPlatformLoggingEnabled = enabled;
+            return (T)this;
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="applicationOptions"></param>
+        /// <returns></returns>
+        public T WithOptions(ApplicationOptions applicationOptions)
```

Comment:
Do we want this to be internal so that the PublicClientApplication / ConfidentialClientApplication have explicit Create() static methods that take the ApplicationOptions in or do we just want them to be able to say "WithOptions()" like the API is defined now?

Reply:
Yup, I already had the static methods (with the appropriate PCA/CCA options classes).  I've now moved WithOptions() to be protected in the abstract base class here so that in order to use them you pass them in at construction time.  And then you could later do a "WithXXX" for a particular property if you want to override the values in the config, or set something that wasn't in the config in the first place.

---
In reply to: [245863273](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/761#discussion_r245863273) [](ancestors = 245863273)